### PR TITLE
fix: for ELECTRON_RUN_AS_NODE=1 not working in Cypress 10

### LINKED ISSUE #23636: Unable to start Cypress with ELECTRON_RUN_AS_NODE=1
### Current behavior

If you set env with `ELECTRON_RUN_AS_NODE=1`, with MacOS, you are not able to start the Cypress with
```bash
ELECTRON_RUN_AS_NODE=1 npx cypress run
```
You'll get error:
`Error: Cannot find module '/Users/<username>/Library/Caches/Cypress/10.6.0/Cypress.app/Contents/MacOS/Contents/Resources/app/index.js'`

Checking the code, it seems that here is where it went wrong:
https://github.com/cypress-io/cypress/blob/b615e8ac30b681b4dc8485afe9814ff453da1fbf/cli/lib/exec/spawn.js#L175

The correct path, at least for Mac, should be:
```javascript
path.join(state.getBinaryPkgPath(path.join(executable, '..', '..', '..')), '..', 'index.js');
```

### Desired behavior

Cypress should start correctly

### Test code to reproduce

Try run with
```bash
ELECTRON_RUN_AS_NODE=1 npx cypress run
```

### Cypress Version

10.6.0

### Node version

16.0

### Operating System

macOS

### Debug Logs

_No response_

### Other

Cypress 10.7.0 (latest) also has this problem

### PR DESCRIPTION
**Pull request by XiaolongLyft**
- Closes #23636

### User facing changelog

Fix for user not able to use `ELECTRON_RUN_AS_NODE=1` to [run Cypress without xvfb](https://docs.cypress.io/guides/continuous-integration/introduction#Xvfb).
During the fix, also found a small bug about incorrect version check for `electron` browser.

### Additional details

Need to include package `electron` in one of packages' dependencies. Otherwise, the package will not be included in the binary's packages.
I included it in `server/package.json`, but probably can move `electron` in `devDependencies` to `dependencies`? @lmiller1990  https://github.com/cypress-io/cypress/blob/d52e610c0036a6c37196883b1c270a2e93ec91ff/package.json#L147

### Steps to test

Now running `ELECTRON_RUN_AS_NODE=1 npx cypress run --browser chrome --headless` will work.
But running `ELECTRON_RUN_AS_NODE=1 npx cypress run` with default `electron` browser still won't work with error:
```
electron_1.BrowserWindow is not a constructor
TypeError: electron_1.BrowserWindow is not a constructor
```
But I'm not sure if this is something this PR should fix.

### How has the user experience changed?

N/A

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: cli/lib/exec/spawn.js [hunk 1]
```diff
@@ -172,7 +172,11 @@ module.exports = {
 
         if (stdioOptions.env.ELECTRON_RUN_AS_NODE) {
           // Since we are running electron as node, we need to add an entry point file.
-          startScriptPath = path.join(state.getBinaryPkgPath(path.dirname(executable)), '..', 'index.js')
+          if (isPlatform('darwin')) {
```
Comment (lmiller1990) [inline cli/lib/exec/spawn.js:175]: Just to clarify if this fixing the issue on darwin only?

Also how did you test this - did you build the project for darwin? Or did you just pick the correct path (no actual testing)?

I've always had problems doing this locally, but I can turn on darwin builds and generate a binary via our CI for testing.

File: packages/server/lib/browsers/utils.ts [hunk 2]
```diff
@@ -249,7 +249,7 @@ const getBrowsers = async () => {
     return browsers
   }
 
-  const version = process.versions.chrome || ''
+  const version = process.versions.electron || ''
```
Comment (lmiller1990) [inline packages/server/lib/browsers/utils.ts]: Hmm this seems like a pretty obvious bug, surprised it flew under the radar for so long. Will need to check this, good job calling this out

File: packages/server/package.json [hunk 3]
```diff
@@ -56,6 +56,7 @@
     "dayjs": "^1.9.3",
     "debug": "^4.3.2",
     "duplexify": "4.1.1",
+    "electron": "21.0.0",
```
Comment (lmiller1990) [inline packages/server/package.json]: I think this is intentionally a devDep - we bundle electron in later during the build step, and it's declared in the top level package.json. I *think* adding this here as a regular dep will bloat the binary with an additional electron (need to check this).

Why do we need to add it here?

**Comment by lmiller1990** [comment]
Nice job, main question is around https://github.com/cypress-io/cypress/pull/24557#discussion_r1014991317

I tried running this is my local Cypress dev env...

```
yarn run v1.22.19
$ cross-env TZ=America/New_York node ../../scripts/cypress run --e2e --project . --browser chrome --headless
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '/Users/lachlan/code/work/cypress/Contents/Resources/app/index.js'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
$ ls /Users/lachlan/code/work/cypress/Contents/Resour
$ ELECTRON_RUN_AS_NODE=1 yarn cypress:run:e2e --browser chrome --headless
```

How did you test this? Darwin only bugfix? Please bit a bit more specific about the bug/environment, and I can give it a test.

If you need a fully built binary for a specific environment(s) please post here, I can update our CI config and it'll built an installed binary for whatever OS you need with all the dependencies installed. You could then add that to your package.json and we can test the patch against some real projects before we ship it to verify it's working as expected. Thanks!!

File: cli/lib/exec/spawn.js (hunk 1 above)
Comment (XiaolongLyft) [reply cli/lib/exec/spawn.js:175]: >Just to clarify if this fixing the issue on darwin only?

I only tested in darwin - it would be nice if we can have a full covered test via CI with Linux and Windows covered.

How I did the test (in MacOS M1):
Basically, I followed the step here: https://github.com/cypress-io/cypress/issues/19908#issuecomment-1131761214
to build a local binary (using `yarn binary-build --platform darwin`) and moved it into my Libaray/Caches, and tested it in one of my real world projects.

File: packages/server/package.json (hunk 3 above)
Comment (XiaolongLyft) [reply packages/server/package.json]: The reason why we're seeing error
```
Cannot find module 'electron'
```
is that in the binary file, in MacOS for example, in `/Library/Caches/Cypress/10.11.0/Cypress.app/Contents/Resources/app/node_modules`, there does not exist the needed `electron` package (you can check in existing released apps), so either we somehow fix it later in the build step to include that package, or we define it in deps.
I do see the CI checks failed because of a larger zip size, so definitely we can look into this more. But either way, the binary size would probably increase.

**Comment by XiaolongLyft** [comment]
Thanks for looking into this really quick @lmiller1990 ! Appreciate that!
>How did you test this? 

I want to mention that this fix is actually in 2 different places, `spawn.js` is in `cli`, and other fixes are suppose to affect the Cypress binary app.
So to effectively test this, I locally built cli and the binary. I replaced the cli section in one of my real projects' in `node_modules`, and I also replaced the binary in `~/Library/Caches/Cypress`. With those 2 parts updated, I can successfully run `ELECTRON_RUN_AS_NODE=1 npx cypress run --browser chrome --headless`.

However, I should point out:
Running `ELECTRON_RUN_AS_NODE=1 npx cypress run` with default `electron` browser still won't work with error:
```
electron_1.BrowserWindow is not a constructor
TypeError: electron_1.BrowserWindow is not a constructor
```
And I think that is not something related to the original issue.

File: packages/server/lib/browsers/utils.ts (hunk 2 above)
Comment (XiaolongLyft) [reply packages/server/lib/browsers/utils.ts]: The fix would not work without this. Otherwise, Cypress would not spin up electron browser with `ELECTRON_RUN_AS_NODE=1` because `version` would be an empty string.

File: cli/lib/exec/spawn.js (hunk 1 above)
Comment (lmiller1990) [reply cli/lib/exec/spawn.js:175]: Right I see, good idea, that's an easy way to test it quickly. You wouldn't even need to build it, you could just modify an existing binary.

Anyway, I updated our config to generate a full, installable binary for MacOS: https://github.com/cypress-io/cypress/compare/lmiller-xiaolong/issue-23636?expand=1

It takes a while (1 hour, CI + build). Once it's completed, a bot will post on that link with the install command, and we can give it a test. Please wait a bit, thanks! I'd like to get this tested and merged by the end of the week.

File: packages/server/lib/browsers/utils.ts (hunk 2 above)
Comment (emilyrohrbough) [reply packages/server/lib/browsers/utils.ts]: I had asked @flotwig about this before. He said had told me:

> The Chrome key there is the version of chromium bundled into Electron, we use that as the display version because it's more useful to the user since it correlates to Chrome versions. Whereas "Electron 19" to a user is less useful.

File: cli/lib/exec/spawn.js (hunk 1 above)
Comment (lmiller1990) [reply cli/lib/exec/spawn.js:175]: Hmm it seems the binary wasn't generated, I might have done something incorrectly. I can take a look next week (Mon 13th Nov).

File: packages/server/lib/browsers/utils.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/server/lib/browsers/utils.ts]: Oh, right, neat trick. We should add a comment.

**Comment by lmiller1990** [comment]
Hey @XiaolongLyft, I am trying to get this to build on MacOS so we can test it out, hopefully this commit will build the binary. https://github.com/cypress-io/cypress/commit/763d1548cd518cbed962c6ad6fd7496e8c030e3f. Doing it on a separate branch since I don't think an external contribute can build the MacOS binaries.

Just letting you know I didn't forget about this, aiming to get it built+tested in the next 24 hours.

**Comment by lmiller1990** [comment]
Ok @XiaolongLyft can you give these a test locally (and CI, if you want) and make sure it's behaving as you expected? *Edit, I think this build is not working, we need to rebase of develop... see below*.

<details><summary>Builds</summary>

[Circle has built](https://circleci.com/gh/cypress-io/cypress/1922776) the darwin arm64 version of the Test Runner.

Learn more about this pre-release platform-specific build at https://on.cypress.io/installing-cypress#Install-pre-release-version.

Run this command to install the pre-release locally:
```
npm install https://cdn.cypress.io/beta/npm/11.1.0/darwin-arm64/lmiller-xiaolong/issue-23636-763d1548cd518cbed962c6ad6fd7496e8c030e3f/cypress.tgz
```

.................

[Circle has built](https://circleci.com/gh/cypress-io/cypress/1922835) the darwin x64 version of the Test Runner.

Learn more about this pre-release platform-specific build at https://on.cypress.io/installing-cypress#Install-pre-release-version.

Run this command to install the pre-release locally:
```
npm install https://cdn.cypress.io/beta/npm/11.1.0/darwin-x64/lmiller-xiaolong/issue-23636-763d1548cd518cbed962c6ad6fd7496e8c030e3f/cypress.tgz
```
______

</details>

Hmm this isn't working for me, I got this in my existing project.

```
$ /Users/lachlan/code/rhythm/node_modules/.bin/cypress run --component
It looks like this is your first time using Cypress: 11.1.0


Cypress failed to start.

This may be due to a missing library or dependency. https://on.cypress.io/required-dependencies

Please refer to the error below for more details.

----------

/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/MacOS/Cypress: bad option: --no-sandbox
/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/MacOS/Cypress: bad option: --smoke-test
/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/MacOS/Cypress: bad option: --ping=998

----------

Platform: darwin-arm64 (21.3.0)
Cypress Version: 11.1.0
```

A new project is even more weird:

```
$ /Users/lachlan/code/dump/cy11/node_modules/.bin/cypress run --browser electron
/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/Resources/app/index.js:2
  throw new Error('global.snapshotResult is not defined. This binary has been built incorrectly.')
  ^

Error: global.snapshotResult is not defined. This binary has been built incorrectly.
    at Object.<anonymous> (/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/Resources/app/index.js:2:9)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1175:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Module._load (node:internal/modules/cjs/loader:829:12)
    at c._load (node:electron/js2c/asar_bundle:5:13339)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47
error Command failed with exit code 1.
```

I think I might have broken something when I rebased Cypress 11 into your branch, let me try again.

___

I'm going to try and pick this up and get it over the line... I force pushed to your branch, rebased on the latest develop. https://github.com/cypress-io/cypress/pull/24557/commits/abb607d25f1af0a98795758a1b2692ace94a3c9e

File: packages/server/lib/browsers/utils.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/server/lib/browsers/utils.ts]: @XiaolongLyft can you please revert this, as pointed out it's intentional.

File: packages/server/package.json (hunk 3 above)
Comment (lmiller1990) [reply packages/server/package.json]: Yep, CI check: https://app.circleci.com/pipelines/github/cypress-io/cypress/45714/workflows/902afa25-43af-4183-bdf4-e3348df80565/jobs/1922692?invite=true#step-113-91

**Comment by lmiller1990** [comment]
Hmm this isn't working for me, I got

```
$ /Users/lachlan/code/rhythm/node_modules/.bin/cypress run --component
It looks like this is your first time using Cypress: 11.1.0


Cypress failed to start.

This may be due to a missing library or dependency. https://on.cypress.io/required-dependencies

Please refer to the error below for more details.

----------

/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/MacOS/Cypress: bad option: --no-sandbox
/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/MacOS/Cypress: bad option: --smoke-test
/Users/lachlan/Library/Caches/Cypress/beta-11.1.0-lmiller-xiaolong/issue-23636-763d1548/Cypress.app/Contents/MacOS/Cypress: bad option: --ping=998

----------

Platform: darwin-arm64 (21.3.0)
Cypress Version: 11.1.0
```

File: packages/server/lib/browsers/utils.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/server/lib/browsers/utils.ts]: Edit: I reverted it for you.

File: packages/server/package.json (hunk 3 above)
Comment (lmiller1990) [reply packages/server/package.json]: I will revert this and see how we go.

**Comment by lmiller1990** [comment]
Going to create a fresh PR and kick this off here https://github.com/cypress-io/cypress/pull/25207
